src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs

```diff
@@ -0,0 +1,243 @@
+                _intMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : intCallback;
+            }
+            else if (measurementCallback is MeasurementCallback<float> floatCallback)
+            {
+                _floatMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : floatCallback;
+            }
+            else if (measurementCallback is MeasurementCallback<double> doubleCallback)
+            {
+                _doubleMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : doubleCallback;
+            }
+            else if (measurementCallback is MeasurementCallback<decimal> decimalCallback)
+            {
+                _decimalMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : decimalCallback;
+            }
+            else if (measurementCallback is MeasurementCallback<short> shortCallback)
+            {
+                _shortMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : shortCallback;
+            }
+            else if (measurementCallback is MeasurementCallback<long> longCallback)
+            {
+                _longMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : longCallback;
+            }
+            else
+            {
+                throw new InvalidOperationException(SR.Format(SR.UnsupportedType, typeof(T)));
+            }
+        }
+
+        /// <summary>
+        /// Enable the listener to start listeneing to instruments measurement recording.
+        /// </summary>
+        public void Start()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (s_allStartedListeners.AddIfNotExist(this, (listener1, listener2) => object.ReferenceEquals(listener1, listener2)))
+            {
+                Meter.NotifyListenerWithAllPublishedInstruments(this);
+            }
+        }
+
+        /// <summary>
+        /// Calls all Observable instruments which the listener is listening to then calls <see cref="SetMeasurementEventCallback" /> with every collected measurement.
+        /// </summary>
+        public void RecordObservableInstruments()
+        {
+            LinkedListNode<Instrument>? current = _enabledMeasurementInstruments.First;
+            while (current is not null)
+            {
+                if (current.Value.IsObservable)
+                {
+                    current.Value.Observe(this);
+                }
+
+                current = current.Next;
+            }
+        }
+
+        /// <summary>
+        /// Disposes the listeners which will stop it from listeneing to any instrument.
+        /// </summary>
+        public void Dispose()
```

Comment:
In the prototype code Dispose() triggers MeasurementComplete notifications for every subscribed instrument and I think we want to preserve that behavior. If we didn't listeners might need to build parallel lists of everything they subscribed to so they could clean it up.

Reply:
agree. I missed that unintentionally.